presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,775 @@
+        this.taskInstanceId = requireNonNull(taskInstanceId, "taskInstanceIs is null");
+        this.outputBuffers = requireNonNull(outputBuffers, "outputBuffers is null");
+        this.state = requireNonNull(state, "state is null");
+        this.tempStorage = requireNonNull(tempStorage, "tempStorage is null");
+        this.threshold = requireNonNull(threshold, "threshold is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        this.finalizerService = requireNonNull(finalizerService, "finalizerService is null");
+        this.finalizerService.addFinalizer(this, this::close);
+
+        tempDataOperationContext = new TempDataOperationContext(Optional.empty(), taskId.getQueryId().toString(), Optional.empty(), new Identity("spooling-buffer", Optional.empty()));
+
+        state.compareAndSet(OPEN, NO_MORE_BUFFERS);
+    }
+
+    @Override
+    public OutputBufferInfo getInfo()
+    {
+        return new OutputBufferInfo(
+                "SPOOLING",
+                state.get(),
+                state.get().canAddBuffers(),
+                state.get().canAddPages(),
+                totalBufferedBytes.get(),
+                totalBufferedPages.get(),
+                totalRowsAdded.get(),
+                totalPagesAdded.get(),
+                ImmutableList.of());
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == BufferState.FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return totalInMemoryBytes.get() / (double) threshold.toBytes();
+    }
+
+    @Override
+    public boolean isOverutilized()
+    {
+        return totalInMemoryBytes.get() > threshold.toBytes();
+    }
+
+    @Override
+    public long getPeakMemoryUsage()
+    {
+        return peakMemoryUsage.get();
+    }
+
+    @Override
+    public ListenableFuture<?> isFull()
+    {
+        return immediateFuture(null);
+    }
+
+    @Override
+    public void addStateChangeListener(StateMachine.StateChangeListener<BufferState> stateChangeListener)
+    {
+        state.addStateChangeListener(stateChangeListener);
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
+    {
+        requireNonNull(newOutputBuffers, "newOutputBuffers is null");
+        checkArgument(outputBuffers.getType() == SPOOLING, "Invalid output buffers type");
+        checkArgument(outputBuffers.isNoMoreBufferIds(), "invalid noMoreBufferIds");
+
+        if (state.get().isTerminal() || outputBuffers.getVersion() >= newOutputBuffers.getVersion()) {
+            return;
+        }
+
+        outputBuffers.checkValidTransition(newOutputBuffers);
+    }
+
+    @Override
+    public void enqueue(Lifespan lifespan, List<SerializedPage> pages)
+    {
+        if (!state.get().canAddPages()) {
+            return;
+        }
+        PendingRead pendingRead;
+        synchronized (this) {
+            this.pages.addAll(pages);
```

Comment:
Alternative: We can use thread safe queue for this.pages and remove  `synchronized (this)`